### PR TITLE
planner,executor: partly fix 'explain for connection' on partition table

### LINKED ISSUE #20019: 'explain for connection' result wrong for partition table
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> show create table xx;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                        |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| xx    | CREATE TABLE `xx` (
  `id` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY HASH( `id` )
PARTITIONS 10 |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select * from xx where id = 1;
Empty set (0.01 sec)

mysql> set @@tidb_partition_prune_mode = 'dynamic-only';
Query OK, 0 rows affected (0.00 sec)

mysql> explain select * from xx where id = 1;
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 10.00    | root      | partition:p1  | data:Selection_6               |
| └─Selection_6           | 10.00    | cop[tikv] |               | eq(test.xx.id, 1)              |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:xx      | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
3 rows in set (0.00 sec)

mysql> select * from xx where id = 1;
Empty set (0.00 sec)

mysql> explain for connection 1;
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 10.00    | root      |               | data:Selection_6               |
| └─Selection_6           | 10.00    | cop[tikv] |               | eq(test.xx.id, 1)              |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:xx      | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
3 rows in set (0.00 sec)
```

### 2. What did you expect to see? (Required)

```
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 10.00    | root      | partition:p1  | data:Selection_6               |
| └─Selection_6           | 10.00    | cop[tikv] |               | eq(test.xx.id, 1)              |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:xx      | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
```

### 3. What did you see instead (Required)

```
+-------------------------+----------+-----------+---------------+--------------------------------+
| id                      | estRows  | task      | access object | operator info                  |
+-------------------------+----------+-----------+---------------+--------------------------------+
| TableReader_7           | 10.00    | root      |               | data:Selection_6               |
| └─Selection_6           | 10.00    | cop[tikv] |               | eq(test.xx.id, 1)              |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:xx      | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+---------------+--------------------------------+
```

Note the access object `partition:p1`, it should be there.

### 4. What is your TiDB version? (Required)

master branch f699bd81a24fafca6f96ea04fb5095bd9047c2ff

```
Release Version: v4.0.0-beta.2-1230-gf699bd81a-dirty
Edition: Community
Git Commit Hash: f699bd81a24fafca6f96ea04fb5095bd9047c2ff
Git Branch: master
UTC Build Time: 2020-09-15 17:15:27
GoVersion: go1.15
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false 
```

This bug is introduced after the partition table refactoring.

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #20019 

Problem Summary:

### What is changed and how it works?

What's Changed:

Do not use cached rows when the `explain for connection` query is on the connection itself.

How it Works:

`explain for connection` uses the cached rows, but the cached rows are not correct for the partition table.
So do not use the cached rows could partly fix the bug.

### Check List 

Tests 

- Unit test

### Release note

- No release note

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @LittleFall  @imtbkcat

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -3166,7 +3166,20 @@ func (b *PlanBuilder) buildExplainFor(explainFor *ast.ExplainForStmt) (Plan, err
...
+		if explainFor.ConnectionID != b.ctx.GetSessionVars().ConnectionID {
+			// Run query in one connection and 'explain for' in another connection, there would be a data race.
+			// So 'explain for' uses cached rows in process info to avoid the data race problem.
+			// See https://github.com/pingcap/tidb/issues/16818
+			//
+			// However, the cached rows are not correct for query on partition table, the
+			// access info of partitions is not set.(Because it requires partition pruning
+			// which is not available then or too expensive to calculate for a second time).
+			//
+			// In the same connection, there is no data race issue, so we ignore the cached rows
```
Comment (SunRunAway) [inline planner/core/planbuilder.go:3178]:
```suggestion
			// In different connections, there is no data race issue, so we ignore the cached rows
```

**Review by imtbkcat — COMMENTED** [review]
LGTM

**Comment by LittleFall** [comment]
LGTM, but seems that this pr cannot completely close [#20019](https://github.com/pingcap/tidb/issues/20019).

File: executor/explainfor_test.go [hunk 2]
```diff
@@ -355,3 +355,34 @@ func (s *testSuite) TestExplainTiFlashSystemTables(c *C) {
...
+	// se1: explain for connection 1
+	tk.MustQuery(fmt.Sprintf("explain for connection %d", connID)).Check(
+		testkit.Rows("TableReader_7 10.00 root partition:p1 data:Selection_6",
+			"└─Selection_6 10.00 cop[tikv]  eq(test.xx.id, 1)",
+			"  └─TableFullScan_5 10000.00 cop[tikv] table:xx keep order:false, stats:pseudo"))
+
+	// This bug is not fixed yet, explain for another connection.
+	// se1: select * from xx where id = 1
+	// se2: explain for connection 1;  -- wrong result
+	tk.MustQuery("select * from xx where id = 1")
```
Comment (zz-jason) [inline executor/explainfor_test.go:379]: how about modifying the filter to see whether the recorded execution plan is changed:

```suggestion
	tk.MustQuery("select * from xx where id = 2")
```

**Review by crazycs520 — COMMENTED** [review]
LGTM, @tiancaiamao Please resolve comment and conflict.

**Comment by tiancaiamao** [comment]
Done @crazycs520

File: executor/explainfor_test.go (hunk 2 above)
Comment (tiancaiamao) [reply executor/explainfor_test.go:517]: It's not used to compare with the previous code.
Just a reminder that the problem is **not completely** fixed:
we can explain for our own connection, but explain for the other connection still get into trouble.

**Comment by tiancaiamao** [comment]
This is not a correct fix. It fixes part of the problem but introduces other problems, so I decide to close it until there is a better solution.

<= data race when `explain for` other connection
`PlanExplainRows` are cached to avoid data race =>

<= when `tidb_partition_prune_mode = 'dynamic-only'` partition pruning is not done in the planning phase
<= so, for partition table, the cached info is incorrect
Do not cached result for the current connection =>
Fix `explain for` get right on current connection =>
But it break the logic for non-partitioned tables!

**Pull request closed without merging**